### PR TITLE
Test change of authority

### LINKED ISSUE #2929: [Bug] WithAuthority override ignored for GetAuthorizationRequestUrl
**Which version of MSAL.NET are you using?**
Latest

**Platform**
.NET 4.7.2

**What authentication flow has the issue?**
* Desktop / Mobile
    * [ ] Interactive
    * [ ] Integrated Windows Authentication
    * [ ] Username Password
    * [ ] Device code flow (browserless)
* Web app 
    * [ ] Authorization code
    * [x] On-Behalf-Of
* Daemon app 
    * [ ] Service to Service calls

Other?

**Is this a new or existing app?**
Existing app new scenario within the app. 

**Repro**
We setup an app to use the "common" endpoint. However, there is a scenario where we want to silently get their consumers* login and we create the following URL:

            Uri authorizationRequestUrl = confidentialClientApplication
                .GetAuthorizationRequestUrl(new List<string> { BaseScope})
                .WithExtraQueryParameters(extraQueryParameters)
                .WithLoginHint(identity.Email) // Note email is primary sign in name (email or phone)
                .WithAuthority(AzureCloudInstance.AzurePublic, "consumers")
                .WithRedirectUri(GetCompleteSignInUrl(authorizationContext.HttpContext.Request.Url.Host).ToString())
                .ExecuteAsync()
                .ConfigureAwait(false)
                .GetAwaiter()
                .GetResult();

Also tried:

            Uri authorizationRequestUrl = confidentialClientApplication
                .GetAuthorizationRequestUrl(new List<string> { BaseScope})
                .WithExtraQueryParameters(extraQueryParameters)
                .WithLoginHint(identity.Email) // Note email is primary sign in name (email or phone)
                .WithAuthority(AadAuthorityAudience.PersonalMicrosoftAccount)
                .WithRedirectUri(GetCompleteSignInUrl(authorizationContext.HttpContext.Request.Url.Host).ToString())
                .ExecuteAsync()
                .ConfigureAwait(false)
                .GetAwaiter()
                .GetResult();

**Expected behavior**
Expected this to generate a URL for the *consumers* endpoint.

**Actual behavior**
Generates URL for the *common* endpoint.

**Possible solution**
Perhaps the options specified here are not being used if provided originally in the app? Maybe support this or throw if it isn't supported.

**Additional context / logs / screenshots**
So the app was originally setup with 'common' via code similar to this:

		ConfidentialClientApplicationOptions applicationOptions;
		applicationOptions = new ConfidentialClientApplicationOptions();
		applicationOptions.ClientId = "fakeId";
		applicationOptions.AadAuthorityAudience = AadAuthorityAudience.AzureAdAndPersonalMicrosoftAccount;
		applicationOptions.Instance = "https://login.microsoftonline.com/common";
		applicationOptions.RedirectUri = "https://example.com";

		var confidentialClientApplicationBuilder = ConfidentialClientApplicationBuilder.CreateWithApplicationOptions(applicationOptions)
			.WithCertificate(someCert);

### PR DESCRIPTION
**Pull request by SameerK-MSFT**
Fixes #2929

**Changes proposed in this request**
Fix handling of authority

**Testing**
Updated tests

**Performance impact**
None

**Documentation**
- [x] All relevant documentation is updated.

### CODE REVIEW

File: src/client/Microsoft.Identity.Client/AppConfig/AuthorityInfo.cs [hunk 1]
```diff
@@ -519,8 +519,8 @@ public static IAuthorityValidator CreateAuthorityValidator(AuthorityInfo authori
                         }
 
                         return updateEnvironment ?
-                                CreateAuthorityWithTenant(CreateAuthorityWithEnvironment(configAuthorityInfo, account.Environment).AuthorityInfo, account?.HomeAccountId?.TenantId) :
-                                CreateAuthorityWithTenant(configAuthorityInfo, account?.HomeAccountId?.TenantId);
+                                CreateAuthorityWithTenant(CreateAuthorityWithEnvironment(requestAuthorityInfo, account.Environment).AuthorityInfo, account?.HomeAccountId?.TenantId) :
```
Comment (bgavrilMS) [inline src/client/Microsoft.Identity.Client/AppConfig/AuthorityInfo.cs:522]: I don't think this is the correct fix.

File: src/client/Microsoft.Identity.Client/AppConfig/AuthorityInfo.cs [hunk 2]
```diff
@@ -519,8 +519,8 @@ public static IAuthorityValidator CreateAuthorityValidator(AuthorityInfo authori
                         }
```
Comment (bgavrilMS) [inline src/client/Microsoft.Identity.Client/AppConfig/AuthorityInfo.cs:519]: I think the better fix is to change line 516 from `IsCommonOrganizationsOrConsumersTenant` to `IsCommonOrOrganizationsTenant`. Smth like

```csharp
if (!requestAuthority.IsCommonOrOrganizations())
{
    return requestAuthority;        
}

if (requestAuthority.IsConsumers())
{
    // let consumers be there.
}
```

File: tests/Microsoft.Identity.Test.Unit/ApiConfigTests/AuthorityTests.cs [hunk 3]
```diff
@@ -245,11 +244,11 @@ public async Task DifferentHostsWithAliasedAuthorityAsync()
             //Checking for aliased authority. Should not throw exception when a developer configures an authority on the application
             //but uses a different authority that is a known alias of the previously configured one.
             //See https://github.com/AzureAD/microsoft-authentication-library-for-dotnet/issues/2736
-            _harness.HttpManager.AddInstanceDiscoveryMockHandler(TestConstants.PrefCacheAuthorityCommonTenant);
+            _harness.HttpManager.AddInstanceDiscoveryMockHandler(TestConstants.AuthorityCommonTenant);
```
Comment (bgavrilMS) [inline tests/Microsoft.Identity.Test.Unit/ApiConfigTests/AuthorityTests.cs:247]: I recommend starting with a test instead.

**Comment by bgavrilMS** [comment]
My proposal is to not accept this. App dev should re-create the app object for this. We should throw an exception or log an error.

File: src/client/Microsoft.Identity.Client/AppConfig/AuthorityInfo.cs (hunk 1 above)
Comment (SameerK-MSFT) [reply src/client/Microsoft.Identity.Client/AppConfig/AuthorityInfo.cs:522]: This gives precedence to request over configuration. We can discuss tomorrow.

**Comment by pmaytak** [comment]
This PR can be closed; #3946 covers the approach we agreed to do?

**Comment by bgavrilMS** [comment]
Closing as not the right solution.

**Pull request closed without merging**